gui/src/vmm/cpu/x86_64.rs

```diff
@@ -32,10 +33,79 @@ impl<H: Hypervisor, S: Screen> Breakpoints for CpuManager<H, S> {
 
 impl<H: Hypervisor, S: Screen> SwBreakpoint for CpuManager<H, S> {
     fn add_sw_breakpoint(&mut self, addr: u64, kind: usize) -> TargetResult<bool, Self> {
-        todo!()
+        if self.sw_breakpoints.contains_key(&addr) {
+            return Ok(false);
+        }
+
+        let breakpoint_size = NonZero::new(kind).unwrap();
+
+        let cpu = self
+            .cpus
+            .first_mut()
+            .ok_or(TargetError::Errno(Self::GDB_ENOENT))?;
+
+        let translated_addr = cpu
+            .debug_mut()
+            .unwrap()
+            .translate_address(addr.try_into().unwrap())
+            .ok_or(TargetError::Errno(Self::GDB_ENOENT))?;
+
+        // Get data.
+        let src = self
+            .hv
+            .ram()
+            .lock(translated_addr, breakpoint_size)
+            .ok_or(TargetError::Errno(Self::GDB_EFAULT))?
+            .as_mut_ptr();
```

Comment:
The `LockedAddr` is dropped at the end of statement. That mean the returned pointer will be invalid.